### DISCUSSION TITLE
How to create an instance of SQLiteDataProviderClassic in a class derived from DataConnection

### DISCUSSION THREAD

**Discussion by alexey-leonovich** [linq2db/linq2db#4435] [Q&A] [answered]
I have a class derived from `DataConnection` and in it's .ctor I call `: base(IDataProvider dataProvider, DbConnection connection)`. But I cannot pass neither `new SQLiteDataProvider(ProviderName.SQLiteClassic)` nor `new SQLiteDataProviderClassic()` (first one is abstract, last one is internal). I had to create a provider derived from `SQLiteDataProvider` (in its .ctor I call `: base(ProviderName.SQLiteClassic)`) - but is that correct? For example `SqlCeDataProvider` is public and in same situation I can create it directly.

**Comment by MaceWindu** [comment, marked as answer]
To get provider instance, use `<DB>Tools.GetDataProvider` method. In your case:

```cs
SQLiteTools.GetDataProvider(ProviderName.SQLiteClassic)
```

**Reply by alexey-leonovich** [reply]
Should I use this method for all providers (e. g. use `SqlCeTools.GetDataProvider()` instead of `new SqlCe.SqlCeDataProvider()`)?
Thank you.

**Reply by MaceWindu** [reply]
Yes. You shouldn't create new provider instances if you don't modify provider itself.
